### PR TITLE
Improve CCL with new algorithms and tests

### PR DESCRIPTION
**Pull request by stal12**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### Pull Request Description

This pull request improves Connected Components Labeling functions as it follows:
1. Introduces the 4-way connectivity version of Spaghetti (Spaghetti4C), generated with the tool described in _Bolelli, F., Allegretti, S., & Grana, C. (2021). One DAG to Rule Them All. IEEE Transactions on Pattern Analysis and Machine Intelligence._
2. Introduces the parallel implementation of Spaghetti and Spaghetti4C algorithms.
3. Improves regression tests introducing the validation of corner cases, like chessboard images, single row and single column images, etc.
4. Changes the default algorithm to the current state of the art.

### CODE REVIEW

File: modules/imgproc/test/test_connectedcomponents.cpp [hunk 1]
```diff
@@ -357,4 +358,302 @@ TEST(Imgproc_ConnectedComponents, spaghetti_bbdt_sauf_stats)
...
+            1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6, 0, 7, 0, 8;
+
+        output_4c.push_back(cv::Mat1b(15, 1));
+        output_4c.back() <<
+            1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6, 0, 7, 0, 8;
+    }
+
+    int ccltype[] = { cv::CCL_DEFAULT, cv::CCL_WU, cv::CCL_GRANA, cv::CCL_BOLELLI, cv::CCL_SAUF, cv::CCL_BBDT, cv::CCL_SPAGHETTI };
+
+    for (size_t i = 0; i < input.size(); ++i) {
```
Comment (alalek) [inline modules/imgproc/test/test_connectedcomponents.cpp:635]: It is better to split scenarios on dedicated tests cases.
It is extremely useful to debug scenario in the middle (through test filtering instead of passing code).

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

**Review by alalek — APPROVED** [review]

**Comment by alalek** [comment]
@stal12 Could you please take a look on this problem: #21366 ?
